RHDEVDOCS-6415: Content creation for Pipelines 1.19 RN (#94698)
openshift/openshift-docs: modules/op-release-notes-1-19.adoc

[id="new-features-1-19_{context}"]
== New features

In addition to fixes and stability improvements, the following sections highlight what is new in {pipelines-title} 1.19:

[id="pipelines-new-features-1-19_{context}"]
=== Pipelines

* With this update, you can now specify custom `securityContext` settings under the `EventListener` YAML specification. When custom `securityContext` is enabled, user-defined values override the default configuration. If no custom values are provided, the default `securityContext` setting is applied automatically.
+
.Example configuration for enabling custom `securityContext` in the `EventListener` YAML spec
[source,yaml]
----
spec:
  serviceAccountName: tekton-triggers-example-sa
  resources:
    kubernetesResource:
      spec:
        template:
          spec:
            securityContext:
              runAsNonRoot: true
            containers:
            - resources:
                requests:
                  memory: "64Mi"
                  cpu: "250m"
                limits:
                  memory: "128Mi"
                  cpu: "500m"
              securityContext:
                readOnlyRootFilesystem: true
----

> * With this update, you can configure custom database credentials for {tekton-results} using the `TektonConfig` custom resource (CR). This eliminates the need to rely on default PostgreSQL secrets that use default usernames and passwords.

Reviewer: Is there an example for how to pull in a secret, or would such an example belong in a piece of documentation separate from the release notes?

Author: Hi @aThorp96, I am not aware of an example that can be used for this. If either you or Shiv Verma can help me with a piece of code, I can accommodate it and add to the content.

Reviewer 2: Updated on google doc

* With this update, {pac} no longer creates a pending status on GitHub Pull Requests when an unauthorized bot user attempts to trigger a `PipelineRun`. Instead of generating a blocking status check, such requests are now silently disallowed.

> * With this update, a new metric, `pipelines_as_code_git_provider_api_request_count`, tracks the number of API calls made by {pac} to Git providers, such as GitHub, GitLab, and Gitea. The metric also helps to monitor API rate limit usage across Git providers, namespaces, event types, and repositories.

Reviewer: ```suggestion
* With this update, a new metric, `pipelines_as_code_git_provider_api_request_count`, tracks the number of API calls made by {pac} to Git providers, such as GitHub, GitLab, and Gitea. The metric also helps to monitor API rate limit usage per Git provider, namespace, event type, and repository.
```

Not sure if my suggestion is preferable or not, but the metric is tagged with the api request's git provider, namespace, etc. My reading of the original wording made me think it would not be tagged with those facets. I'm not an expert in technical writing though; if the original wording is appropriate then it is fine. I just wanted to make sure it accounted for the tags.

Reviewer 2: updated on google doc

Author: Fixed.

> * With this update, repository URLs are now validated during creation to ensure they are properly formatted and use valid schemes, such as `http` or `https`. This enhancement helps prevent misconfigurations and runtime errors.

Reviewer: ```suggestion
* With this update, Repository URLs are now validated during creation to ensure they are properly formatted and use valid schemes, such as `http` or `https`. This enhancement helps prevent misconfigurations and runtime errors.
```
Capitalizing `Repository` to disambiguate between a code repository and a Pipelines as Code Repository Custom Resource

Reviewer: It looks like k8s resources are generally not capitalized in the rest of the document. If there is another way to disambiguate this that would be useful. There's no other context in this bullet to indicate this is related to Pipelines as Code, and in my experience confusing and conflating a git repository and a PaC Repository CR is very common when talking with users.

* With this update, `PipelineRun` status comments now render correctly in markdown on Bitbucket Data Center and Bitbucket Cloud, instead of appearing as raw strings in the pull request UI.

* With this update, when the `generateSigningSecret` key is enabled, the {pipelines-title} Operator automatically creates Cosign key pairs, specifically `cosign.key` (private key) and `cosign.pub` (public key). These key pairs are then stored within the `signing-secrets` secret, located in the `tekton-pipelines` namespace. You can enable the `generateSigningSecret` key by setting it to `true` in the `TektonConfig` custom resource.
+
.Example of enabling Cosign key pairs
[source,yaml]
----
apiVersion: operator.tekton.dev/v1
kind: TektonConfig
metadata:
  name: config
spec:
  chain:
    disabled: false
    generateSigningSecret: true
----

> * With this update, users can retrieve logs from Splunk that were forwarded by OpenShift Logging. To configure this feature, you can set environment variables such as `SPLUNK\_SEARCH\_TOKEN, LOGGING\_PLUGIN\_QUERY\_PARAMS` (including index details such as index=konflux), and `LOGGING\_PLUGIN\_API\_URL` (with the Splunk endpoint and port number).

Reviewer: Is this through Tekton Results? On what resource do the environment variables need to be set?

Reviewer 2: yes this is through tekton results

Reviewer 2: updated on the google doc

Author: Fixed.

> * With this update, the {tekton-results} watcher leverages `StatefulSet` ordinals for improved high availability and workload distribution as an alternative to leader election.

Reviewer: the release notes say this is Tekton Chains but the example config is inside of the `chains` part of the Tekton Config. Is that correct?

Also, do i understand right that Tekton Results is part of a stateful set instead of a deployment? @khrm is that correct or is this a setting which can be enabled?

Reviewer 2: @aThorp96 can you clarify your ask ?

Reviewer 3: Probably, what's meant here is that we can set up run watcher (knative controller) as a statefulset with multiple replica.

+
--
.Example enabling the StatefulSet ordinals for the {tekton-results} watcher
[source,yaml]
----
spec:
  chains:
    performance:
      disable-ha: false
      buckets: 4
      replicas: 4
      statefulset-ordinals: true
----

:FeatureName: Using `StatefulSet` ordinals for high availability
include::snippets/technology-preview.adoc[]
--

> * With this update, the {tekton-results} API supports response filtering to minimize payload size and enhance network efficiency. This update enables users to define the specific fields included in API responses, which significantly benefits List operations by preventing the retrieval of entire objects, thus optimizing performance.

Reviewer: ```suggestion
* With this update, the {tekton-results} API supports response field filtering to minimize payload size and enhance network efficiency. This update enables users to define the specific fields included in API responses, which significantly benefits List operations by preventing the retrieval of entire objects, thus optimizing response latency and IO performance.
```

Author: Fixed.

* With this update, tasks in the web console now feature `displayName` properties, providing clearer, more user-friendly names within the {pipelines-title} interface. This behavior replaces the use of technical task names and brings the `displayName` parameter functionality, previously found in TektonDashboard, to the web console.

> * With this update, {pac} extends GitOps workflow support in GitLab by enabling GitOps comments on pushed commits. This enhancement allows users to receive pipeline status updates directly on commit events, in addition to merge requests. Previously, this feature was only available for GitHub. # (Need to confirm if this is added in Google Docs)

Reviewer: > `# (Need to confirm if this is added in Google Docs)`

I'm not an asciidoc person so want to confirm this is not left in by mistake 😛

Author: Fixed.

* With this update, parameter naming conventions across the `StepActions` feature have been unified for consistency. The updated `git-clone` `StepActions` uses consistent casing with `cache-fetch` and `cache-upload` `StepActions`.

* With this update, the `/ok-to-test` memory feature is disabled by default. This precaution helps mitigate the risk of malicious code execution within testing environments. # (Need to confirm if this is added in Google Docs)

* With this update, support is provided for expanding dynamic variables from remote pipeline definitions. This enhancement improves pipeline composition capabilities, which enables additional flexible pipeline architectures.

* With this update, the {tekton-results} Watcher retries reconciliation multiple times before removing the finalizer, as long as the `storedDeadline` duration has not been exceeded. This improves reliability by reducing the chances of missing `TaskRun` or `PipelineRun` storage.

> * With this update, {tekton-results} supports field-based filtering for `Get` and `List` requests on `TaskRuns` and `PipelineRun`. When filter fields are specified, only the requested data is returned, improving response times and reducing network load.

Reviewer: I believe this duplicates the bullet point on line 90:
> `* With this update, the {tekton-results} API supports response filtering to minimize payload size and enhance network efficiency. This update enables users to define the specific fields included in API responses, which significantly benefits List operations by preventing the retrieval of entire objects, thus optimizing performance.`

Reviewer 2: let's go with the one in line 90 then

* With this update, you no longer need to switch to Kubernetes APIs to manage `ManualApprovalGate` Tasks. You can now use the `opc` tool to `approve`, `reject`, or `inspect` gates directly with commands such as `opc mag approve --task <name>` or `opc mag list`.

* With this update, the `--live` flag in the `opc version --live` command queries the cluster to display the running versions of Chains, Pipelines, and other Tekton controllers, not just the local binary versions. In the previous release, users had to manually inspect cluster resources to determine which controller versions were deployed. Now, that information is displayed by using this command.

* With this update, the `opc version` command now displays the {tekton-results} API client library version in addition to {pac} and Tekton versions. Before this release, this command only displayed versions for `opc`, `pac`, and `tkn`. This enhancement helps operators to easily verify cross-component compatibility between CLI tools and the cluster.

> * With this update, the `remember-ok-to-test` value in {tekton-config} is set to `false` by default to reduce the risk of executing untrusted or malicious code in test environments.

Reviewer: Should it be specified that this is under the `pipelines-as-code` section of the tekton-config?

> * With this update, {pac} now supports triggering `PipelineRun` in GitLab using GitOps style comments such as `/test` and `/retest` on pushed commits.

Reviewer: I think this duplicates the item on line 94: 
> `* With this update, {pac} extends GitOps workflow support in GitLab by enabling GitOps comments on pushed commits. This enhancement allows users to receive pipeline status updates directly on commit events, in addition to merge requests. Previously, this feature was only available for GitHub.`

Author: Removed from the content as discussed.

// This module is included in the following assemblies:
// * release_notes/op-release-notes-1-19.adoc

:_mod-docs-content-type: REFERENCE
[id="op-release-notes-1-19_{context}"]
= Release notes for {pipelines-title} General Availability 1.19

With this update, {pipelines-title} General Availability (GA) 1.19 is available on {OCP} 4.15 and later versions.

[id="new-features-1-19_{context}"]
== New features

In addition to fixes and stability improvements, the following sections highlight what is new in {pipelines-title} 1.19:

[id="pipelines-new-features-1-19_{context}"]
=== Pipelines

> * With this update, you can now specify custom `securityContext` settings under the `EventListener` YAML specification. When custom `securityContext` is enabled, user-defined values override the default configuration. If no custom values are provided, the default `securityContext` setting is applied automatically.

Author: Fixed.

+

> .Example configuration for enabling custom `securityContext` in the `EventListener` YAML spec

Author: Fixed.

[source,yaml]
----

> spec:

Author: Fixed.

  serviceAccountName: tekton-triggers-example-sa
  resources:
    kubernetesResource:
      spec:
        template:
          spec:
            securityContext:
              runAsNonRoot: true
            containers:
            - resources:
                requests:
                  memory: "64Mi"
                  cpu: "250m"
                limits:
                  memory: "128Mi"
                  cpu: "500m"
              securityContext:
                readOnlyRootFilesystem: true

> ----

Author: Fixed.

> * With this update, you can configure custom database credentials for {tekton-results} using the `TektonConfig` custom resource (CR). This eliminates the need to rely on default PostgreSQL secrets that use default usernames and passwords.

Author: Fixed.

> * With this update, you can configure custom database credentials for {tekton-results} using the `TektonConfig` custom resource (CR). This eliminates the need to rely on default PostgreSQL secrets that use default usernames and passwords.

Author: Hey @eromanova97, we have incorporated the changes from this file https://github.com/tektoncd/operator/pull/2724/files#diff-57112ca6edd61727c1d2186f06ea5f1b25084c7a85fbc3650386fb08d4e054cf[docs/TektonResult.md] file. It states that we use TektonResult as kind.

> * With this update, you can configure custom database credentials for {tekton-results} using the `TektonConfig` custom resource (CR). This eliminates the need to rely on default PostgreSQL secrets that use default usernames and passwords.

Author: Discussed it with @waveywaves before coming to this conclusion.

+
.Example for adding custom database credentials for {tekton-results}
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonResult
metadata:
  name: result
spec:
  db_secret_name: # optional: custom database secret name
  db_secret_user_key: # optional
  db_secret_password_key: # optional
...
----

> * With this update, {pac} no longer creates a `Pending` status on GitHub Pull Requests when an unauthorized bot user attempts to trigger a `PipelineRun`. Instead of generating a blocking status check, such requests are now silently disallowed.

Author: Fixed.

* With this update, a new metric, `pipelines_as_code_git_provider_api_request_count`, tracks the number of API calls made by {pac} to Git providers, such as GitHub, GitLab, and Gitea. The metric also helps to monitor API rate limit usage per Git provider, namespace, event type, and repository.

* With this update, URLs in the `Repository` CR are now validated during creation to ensure they are properly formatted and use valid schemes, such as http or https. This enhancement helps prevent misconfigurations and runtime errors.

* With this update, `PipelineRun` status comments now render correctly in markdown on Bitbucket Data Center and Bitbucket Cloud, instead of appearing as raw strings in the pull request UI.

> * With this update, when the `generateSigningSecret` key is enabled, the {pipelines-title} Operator automatically creates Cosign key pairs, specifically `cosign.key` (private key) and `cosign.pub` (public key). These key pairs are then stored within the `signing-secrets` secret, located in the `tekton-pipelines` namespace. You can enable the `generateSigningSecret` key by setting it to `true` in the `TektonConfig` custom resource.

Author: Fixed.

> * With this update, when the `generateSigningSecret` key is enabled, the {pipelines-title} Operator automatically creates Cosign key pairs, specifically `cosign.key` (private key) and `cosign.pub` (public key). These key pairs are then stored within the `signing-secrets` secret, located in the `tekton-pipelines` namespace. You can enable the `generateSigningSecret` key by setting it to `true` in the `TektonConfig` custom resource.

Author: Fixed.

+

> .Example of enabling Cosign key pairs

Author: Fixed.

[source,yaml]
----
apiVersion: operator.tekton.dev/v1
kind: TektonConfig
metadata:
  name: config
spec:
  chain:
    disabled: false
    generateSigningSecret: true

> ----

Author: Fixed.

> * With this update, {tekton-results} users can retrieve logs from Splunk that were forwarded by OpenShift Logging. To configure this, the following environment variables must be set:

Author: Discussed it further with @waveywaves and he suggested that we configure this in {tekton-results} deployment.

> * With this update, {tekton-results} users can retrieve logs from Splunk that were forwarded by OpenShift Logging. To configure this, the following environment variables must be set:

Author: Can be added to the {tekton-results} API deployment and adding the example provided by @khrm.

** SPLUNK\_SEARCH\_TOKEN, LOGGING\_PLUGIN\_QUERY\_PARAMS - Including index details such as `index=konflux`
** LOGGING\_PLUGIN\_API\_URL - With the Splunk endpoint and port number

* With this update, the {tekton-results} watcher leverages `StatefulSet` ordinals for improved high availability and workload distribution as an alternative to leader election.
+
--
.Example for enabling `StatefulSet` ordinals for the {tekton-results} watcher
[source,yaml]

> ----

Author: Fixed.

spec:

>   chains:

Author: Fixed.

    performance:
      disable-ha: false
      buckets: 4
      replicas: 4
      statefulset-ordinals: true

> ----

Author: Fixed.

:FeatureName: Using `StatefulSet` ordinals for high availability
include::snippets/technology-preview.adoc[]
--

> * With this update, the {tekton-results} API supports response field filtering or partial responses to minimize payload size and enhance network efficiency. This update enables users to define the specific fields included in API responses, which significantly benefits `List` operations by preventing the retrieval of entire objects, thus optimizing response latency and IO performance.

Author: Fixed.

> * With this update, the {tekton-results} API supports response field filtering or partial responses to minimize payload size and enhance network efficiency. This update enables users to define the specific fields included in API responses, which significantly benefits `List` operations by preventing the retrieval of entire objects, thus optimizing response latency and IO performance.

Author: Fixed.

> * With this update, tasks in the web console now feature `displayName` properties, providing clearer, more user-friendly names within the {pipelines-title} interface. This behavior replaces the use of technical task names and brings the `displayName` parameter functionality, previously found in TektonDashboard, to the web console.

Author: Fixed.

* With this update, the {pipelines-title} entrypoint image includes support for FIPS compliance. The image is now built with appropriate cryptographic dependencies and configurations to meet FIPS requirements, ensuring compatibility in environments with strict security standards.

* With this update, FIPS compliance is enhanced by adding `disable_spire` and `disable_tls` build tags to the entrypoint command compilation process. This enhancement ensures cryptographic symbols are properly removed from the entrypoint binary to meet compliance requirements.

> * With this update, parameter naming conventions across the `StepActions` feature have been unified for consistency. The updated `git-clone` `StepActions` uses consistent casing with `cache-fetch` and `cache-upload` `StepActions`.

Author: Fixed.

* With this update, the `/ok-to-test` memory feature is disabled by default. This precaution helps mitigate the risk of malicious code execution within testing environments.

* With this update, support is provided for expanding dynamic variables from remote pipeline definitions. This enhancement improves pipeline composition capabilities.

> * With this update, the {tekton-results} Watcher retries reconciliation multiple times before removing the finalizer, as long as the `storedDeadline` duration has not been exceeded. This improves reliability by reducing the chances of missing `TaskRun` or `PipelineRun` storage.

Author: Fixed.

* With this update, the Git resolver used in `RemoteResolution` is updated to use the native git binary instead of the pure Go `go-git` library. This change significantly reduces memory consumption and improves clone performance, especially for large repositories. This enhancement uses shallow-clone flags, for example `--depth 1`, to minimize resource usage. No changes to pipeline manifests are required. This enhancement is available starting with version 1.0.0.

* With this update, the `onError` field in {pipelines-title} is enhanced to support Tekton parameter substitution. Previously, the `onError` filed only accepted literal values such as `continue`, `stop`, and `finally`. Starting with version 1.0.0, you can now use substitution tokens, such as `$(params.strategy)` to dynamically determine failure handling behavior at runtime. This allows a single Pipeline definition to flexibly adapt its `onError` policy based on parameters, context, or results.

> * With this update, retry timings for OCI bundle lookups are no longer fixed. New config map keys under the `bundle.resolver.backoff.*` prefix allow cluster administrators to configure the initial retry delay, backoff factor, and maximum retry duration. This helps reduce load on busy registries by preventing aggressive retry behavior. To adjust these settings, update the `config-resolver-bundle` config map.

Author: Fixed.

> * With this update, retry timings for OCI bundle lookups are no longer fixed. New config map keys under the `bundle.resolver.backoff.*` prefix allow cluster administrators to configure the initial retry delay, backoff factor, and maximum retry duration. This helps reduce load on busy registries by preventing aggressive retry behavior. To adjust these settings, update the `config-resolver-bundle` config map.

Author: Fixed.

> * With this release, `StepActions` have been updated from alpha to stable and are now enabled by default. The  `enable-step-actions` flag used in the earlier version is not used and will be removed in a future release. This means workflows can use `StepActions` without requiring feature flags, as the capability is permanently enabled across all clusters.

Author: Fixed.

> * With this update, the Git resolver can now use personal access tokens to authenticate with GitHub or GitLab, avoiding rate limits associated with anonymous API usage. To enable this, add a `token:` field to your Git resolver specification. Tekton automatically injects the token as an HTTP header during resolution to reduce the risk of quota-related errors during remote resolution.

Author: Fixed.

> * With this update, the Git resolver can now use personal access tokens to authenticate with GitHub or GitLab, avoiding rate limits associated with anonymous API usage. To enable this, add a `token:` field to your Git resolver specification. Tekton automatically injects the token as an HTTP header during resolution to reduce the risk of quota-related errors during remote resolution.

Author: @waveywaves will help us with an example.

> * With this update, the Git resolver can now use personal access tokens to authenticate with GitHub or GitLab, avoiding rate limits associated with anonymous API usage. To enable this, add a `token:` field to your Git resolver specification. Tekton automatically injects the token as an HTTP header during resolution to reduce the risk of quota-related errors during remote resolution.

Author: Added a sample in the code and changed the content a bit as well.

> * With this update, the Pipeline scheduler now correctly evaluates result references in fan-out/fan-in patterns. Previously, such Pipelines could fail unpredictably when matrix Tasks relied on result refs.

Author: Fixed.

* With this update, the `opc version` command now displays the {tekton-results} API client library version in addition to {pac} and Tekton versions. Before this release, this command only displayed versions for `opc`, `pac`, and `tkn`. This enhancement helps operators to easily verify cross-component compatibility between CLI tools and the cluster.

> * With this update, the `remember-ok-to-test` value in {tekton-config} is set to `false` by default to reduce the risk of executing untrusted or malicious code in test environments.

Author: Fixed.

> * With this update, build caching is no longer limited to predefined paths for `Go` and `Maven`. A new `--buildCommand <dir>` flag allows you to snapshot and cache any directory, enabling support for additional build systems like `Rust (target/)`, `Gradle (.gradle/)`, or custom toolchains.

Author: Fixed.

> * With this update, Tekton Caches can now automatically detect and use common cache directories when the `--path` flag is not specified. The system inspects the workspace and applies sensible defaults such as `$GOMODCACHE`, `~/.npm`, and `~/.m2`.

Author: Fixed.

> * With this update, Tekton Caches can now automatically detect and use common cache directories when the `--path` flag is not specified. The system inspects the workspace and applies sensible defaults such as `$GOMODCACHE`, `~/.npm`, and `~/.m2`.

Author: Fixed.

> * With this update, Tekton Caches can now automatically detect and use common cache directories when the `--path` flag is not specified. The system inspects the workspace and applies sensible defaults such as `$GOMODCACHE`, `~/.npm`, and `~/.m2`.

Author: Fixed.

> * With this update, Tekton Caches can now be pushed to and retrieved from Google Cloud Storage (GCS) buckets, in addition to existing OCI registry support. To enable this, set the cache backend to a `gcs://bucket/path` URI.

Author: Fixed.

> * With this update, running on Google Kubernetes Engine (GKE) with Workload Identity Federation (WIF) no longer requires embedding key files in Tasks. You can now mount projected volume tokens instead, eliminating the need for long-lived credentials and improving security.

Author: Fixed.

> * With this update, the CLI and Task components now use a consistent environment variable for provider credentials. Setting the environment variable to `CRED_STORE=/tekton/creds/creds.json` now works uniformly across all supported backends, including OCI, GCS, and S3.

Author: Fixed.

+
.Example for enabling the `CRED_STORE=/tekton/creds/creds.json` environment variable
[source,yaml]
----

> containers:

Author: Fixed it as per discussion with @waveywaves.

- name: cache
  env:
  - name: CRED_STORE
    value: /tekton/creds/creds.json
  volumeMounts:
  - name: creds
    mountPath: /tekton/creds
----

> * With this update, cache archives can now be stored in any S3 compatible bucket, including on-premises solutions like MinIO or cloud providers like AWS. To use this feature, specify a URL, such as `s3://my-bucket/cache` as the cache backend.

Author: Fixed.

> * With this update, cache archives are compressed using `Gzip` before being uploaded. This reduces object storage costs and speeds up data transfer, especially for large caches such as `Go` modules.

Author: Fixed.

> * With this update, the code paths for GCS and S3 backends have been unified using the `gocloud.dev` library. This abstraction makes it easier to support additional storage providers, such as Azure Blob Storage or local filesystems.

Author: Fixed.

* With this update, the `fetch` command has been improved to automatically create the destination folder if it does not exist in a new workspace. Previously, the command would fail in such cases, requiring you to create a directory manually.

> * With this update, restored caches retain their original file mode bits, ensuring that executable scripts and other permission-sensitive files function correctly. Previously, restored files default to 0600 permissions, which can prevent scripts from running as expected.

Author: Fixed.

> * With this update, restored caches retain their original file mode bits, ensuring that executable scripts and other permission-sensitive files function correctly. Previously, restored files default to 0600 permissions, which can prevent scripts from running as expected.

Author: Fixed.

* With this update, registry authentication is no longer limited to the default path at `/tekton/home/.docker/config.json`. You can now mount any Docker configuration file and specify its location using the `dockerConfig` parameter in your Task.
+
.Example for enabling the `dockerConfig` parameter in a task
[source,yaml]
----
params:
- name: dockerConfig
  value: /workspace/creds/config.json

> ----

Author: Fixed.

+
.Example for enabling the `SQL_LOG_LEVEL` environment variable
[source,yaml]
----
results:
  options:
    deployments:
       tekton-results-api:
         spec:
           template:
             spec:
               containers:
               - name: api
                 env:
                 - name: SQL_LOG_LEVEL
                   value: debug
----

[id="operator-new-features-1-19_{context}"]
=== Operator

* {pipelines-title} 1.19 introduces a new high availability option for the {tekton-chains} controller, utilizing `StatefulSet` ordinals. This alternative to leader election aims to improve workload balance. While leader election offers rapid failover capabilities, it can sometimes cause uneven distribution of tasks. The implementation of `StatefulSet` ordinals ensures a more consistent distribution of signing keys across replicas, leading to a more balanced workload. Cluster administrators now have the flexibility to choose between prioritizing fast recovery using leader election or achieving a more uniform workload distribution with `StatefulSet` ordinals.
+
--
.Example of enabling the StatefulSet ordinals for the the `Chains` controller
[source,yaml]
----

> spec:

Author: Fixed.

  chains:
    performance:
      disable-ha: false
      buckets: 4
      replicas: 4
      statefulset-ordinals: true
----
:FeatureName: Using `StatefulSet` ordinals for high availability
include::snippets/technology-preview.adoc[]
--

> * With this update, the event-based pruner can now be enabled or disabled through the `TektonConfig` CR by setting the `spec.tektonpruner.disabled` parameter to `false`. Fine-grained configuration is not yet available through the `TektonConfig` CR and should be managed using config maps.

Reviewer 4: The event based pruner is tech preview and this is why it is managed using config maps. Please ensure that the event-bbased pruner is clearly marked tech preview in RNs using the standard snippet.

> * With this update, the event-based pruner can now be enabled or disabled through the `TektonConfig` CR by setting the `spec.tektonpruner.disabled` parameter to `false`. Fine-grained configuration is not yet available through the `TektonConfig` CR and should be managed using config maps.

Author: Fixed.

+

> [NOTE]

Author: Fixed.

====
The existing job-based pruner must be disabled before enabling the event-based pruner.
====

[id="pac-new-features-1-19_{context}"]
=== {pac}

> * Before this update, YAML syntax errors in `PipelineRun` were only reported in logs and Kubernetes events, making them difficult to detect and troubleshoot. With this release, {pac} now comments directly on pull requests when `PipelineRun` YAML validation errors occur. This improves error visibility and simplifies troubleshooting on GitHub, GitLab, and Gitea providers.

Author: Fixed.

> * Before this update, YAML syntax errors in `PipelineRun` were only reported in logs and Kubernetes events, making them difficult to detect and troubleshoot. With this release, {pac} now comments directly on pull requests when `PipelineRun` YAML validation errors occur. This improves error visibility and simplifies troubleshooting on GitHub, GitLab, and Gitea providers.

Author: Yeah, it looks like a bug fix. I will move the content to that section.

> * Before this update, in the GitLab integration, {pac} posted comments on merge requests at the start and end of each `PipelineRun`. This behavior led to excessive comments appearing on merge requests when multiple pipelineruns were triggered. With this release, you can now disable all GitLab comments by setting `disable_all` to `true` in the Repository custom resource (CR).

Author: Fixed.

+
.Example of enabling the Repository CR
[source,yaml]
----
---
apiVersion: "pipelinesascode.tekton.dev/v1alpha1"
kind: Repository
metadata:
  name: test-pac
spec:
  # other fields 
  settings:
    gitlab:
      comment_strategy: "disable_all"
----

> * With this release, {pac} introduces a new metric `pipelines_as_code_git_provider_api_request_count`. This metric tracks the number of API requests made by {pac} to a Git provider in response to an event.

Author: Fixed.

> * With this release, the `TektonConfig` custom resource provides support for two new fields to enable the `cancel-in-progress` feature for pipelineruns in {pac} globally:

Author: Fixed.

** `enable-cancel-in-progress-on-pull-requests`
** `enable-cancel-in-progress-on-push`
+

> When set to true, these fields automatically cancel any in-progress pipelinerun triggered by pull request or push events when there is a new commit. By default, both these fields are set to false.

Author: Fixed.

+
[NOTE]
====
If a PipelineRun includes the `pipelinesascode.tekton.dev/cancel-in-progress` annotation, it will override the corresponding `TektonConfig` setting.
====
+

> .Example configuration for enabling “enable-cancel-in-progress-on-pull-requests” and “enable-cancel-in-progress-on-push” in `TektonConfig` CR

Author: Fixed.

[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
....
platforms:
    openshift:
      pipelinesAsCode:
        additionalPACControllers:
          <controllerName>:
            enable: true
            configMapName:
            secretName:
            settings:
        enable: true
        settings:
          ........
          error-log-snippet: "true"
          enable-cancel-in-progress-on-pull-requests: "false"
          enable-cancel-in-progress-on-push: "false"
          hub-catalog-name: tekton
          hub-url: https://api.hub.tekton.dev/v1
          skip-push-event-for-pr-commits: "true"

>           .......

Author: Fixed.

> ----

Author: Fixed.

* With this release, {pac} supports a new dynamic variable `git_tag`. This variable is used during tag push events and reflects the value of the Git tag. For example, if the tag `v1.0` is pushed to the `Repository` CR, the `git_tag` variable holds the value `v1.0`.
+
.Example configuration for git_tag
[source,yaml]
----
---
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: pull-pr-3
  annotations:
    pipelinesascode.tekton.dev/on-event: ["push"]
    pipelinesascode.tekton.dev/on-target-branch: ["refs/tags/*"]
spec:
  params:
    - name: tag
      value: "{{ git_tag }}"
  pipelineSpec:
    tasks:
        ...
        taskSpec:
          steps:
            ...

>               script: |

Author: Fixed.

                echo "$(params.tag)" # output will be `v1.0`
                exit 0

> ----

Author: Fixed.

> * With this release, a new field, `skip-push-event-for-pr-commits`, has been added to the `TektonConfig` CR. When enabled, {pac} does not trigger pipelineruns for push events if the commit SHA is included in an open pull request. This prevents duplicate pipelineruns for the same commit. By default, this field is set to `true`.

Author: Fixed.

+
.Example configuration for `skip-push-event-for-pr-commits` in `TektonConfig`
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config

> ....

Author: Fixed.

platforms:
    openshift:
      pipelinesAsCode:
        additionalPACControllers:
          <controllerName>:
            enable: true
            configMapName:
            secretName:
            settings:
        enable: true
        settings:
          ........
          hub-url: https://api.hub.tekton.dev/v1
          skip-push-event-for-pr-commits: "true"
          remote-tasks: "true"
          secret-auto-create: "true"
          .......
----

> * With this release, an OpenAPI schema is now integrated into {pac} for the Repository CR. This schema enables IDE autocompletion for Repository CR writing and allows repository explanations via the `kubectl explain` command.

Author: Fixed.

* With this update, when you set the `on-cel-expression` and `on-event` or `on-target-branch` annotations in a repository, the `on-cel-expression` annotation takes precedence. The `on-event` and `on-target-branch` annotations are ignored in this case. To alert users, a warning log and Kubernetes event are generated to indicate this behavior.

[id="pruner-new-features-1-19_{context}"]
=== Pruner

* With this update, the `Pruner` component introduces automated cleanup of `PipelineRun` and `TaskRun` resources in Tekton. It supports the following features:

** Time-based pruning (TTL) - Automatically deletes completed `PipelineRun` and `TaskRuns` after a specified duration. This is controlled using the `ttlSecondsAfterFinished` setting.

** History-based pruning - Retains a limited number of successful and failed runs. Configured using:

*** successfulHistoryLimit
*** failedHistoryLimit
*** historyLimit 

** Flexible configuration levels:

*** Global level: Applies to all namespaces except those prefixed with `kube-` and `openshift-`. 

> *** Namespace level: Applies to all resources in a specific namespace.

Author: Fixed.

:FeatureName: The Pruner component
include::snippets/technology-preview.adoc[]

+
[NOTE]
====

> In this release, only `Global` and `Namespace-level` configurations are available.

Author: I will confirm these with the developers.

====

[id="breaking-changes-1-19_{context}"]
== Breaking changes

> * With this release, the `hub clustertask` command is removed from the CLI, as `ClusterTask` is no longer available on {tekton-hub}. This change eliminates a non-functional entry point and helps reduce user confusion.

Author: Fixed.

* With this release, support for `ClusterTask` objects is removed following their deprecation upstream in {pipelines-title} v1.0.0. As a result, the `tkn clustertask` and `tkn task create` commands from `clustertask` are no longer available.

[id="known-issues-1-19_{context}"]
== Known issues

> * Currently, the event-based pruner does not strictly validate the contents of the `tekton-pruner-default-spec` config map. If invalid configuration keys or malformed values with incorrect field names are provided, the configuration is ignored. As a result, the pruner may fall back to default behavior or skip pruning altogether.

Reviewer 4: The event-based pruner is tech preview and does use config maps for now. I suppest adding a standard tech preview snippet here about the event-based pruner.

> * Currently, the event-based pruner does not strictly validate the contents of the `tekton-pruner-default-spec` config map. If invalid configuration keys or malformed values with incorrect field names are provided, the configuration is ignored. As a result, the pruner may fall back to default behavior or skip pruning altogether.

Author: Fixed.

* Currently, the `--host`, `--token`, `--api-path`, and `--insecure-skip-tls-verify` flags in the `opc results` command are non-functional. Parameters passed through these flags are ignored.

* Currently, the `results` CLI does not return helpful or actionable error messages when there is a configuration error, making it difficult to troubleshoot issues.

* Currently, the `--all-namespaces` flag is non-functional and does not work with the `opc results tr list` and `opc results pr list` commands due to an issue in the endpoints.

[id="fixed-issues-1-19_{context}"]
== Fixed issues

> * Before this update, the `s2i-java` task failed with an error message, `/usr/libexec/s2i/assemble: No such file or directory`. This error occurred due to incorrect script path references. With this update, the default script path for the `s2i-java` task is modified to `/usr/local/s2i`. Other S2i tasks, such as those for Go or .NET, continue to use the `/usr/libexec/s2i/assemble` script path.

Author: Fixed.

* Before this update, logs retrieved from the Amazon Web Services (AWS) S3 bucket were displayed in a random order, making debugging and troubleshooting difficult. With this update, logs from AWS S3 are now correctly ordered chronologically, improving readability and the overall debugging experience.

* Before this update, {pac} required all custom parameters defined in a `Repository` CR to include predefined values. With this update, custom parameters can now be defined without specifying default values in the `Repository` CR. This change enables values to be supplied through webhook payloads and preserves backward compatibility.

* Before this update, when using the `github-push` ClusterTriggerBinding, the `git-clone` command could fail with `HTTP 403` errors. This issue occurred because the `$(body.repository.url)` parameter pointed to the GitHub API URL instead of a valid Git clone URL. With this update, a new parameter, `git-repo-clone-url`, uses `$(body.repository.html_url)` to ensure that the correct repository URL is used for cloning.

> * Before this update, the `buildah` task failed to process build arguments that contained spaces when used with the cluster resolver. This issue affected users migrating from deprecated the `ClusterTask` custom resource (CR). With this update, the `BUILD_ARGS` parameter in the `buildah` task now correctly supports arguments with spaces, for example, `FOO="abc def"`, restoring compatibility with previous functionality.

Author: Fixed.

* Before this update, the *PipelineRun details* page failed to load correctly, preventing users from viewing pipeline execution details. With this update, the loading mechanism is fixed, and `PipelineRun` information is now displayed correctly in the Web UI.

* Before this update, the console plugin styling was outdated due to the upgrade to PatternFly 6 and the removal of deprecated `co-` classes. This caused alignment and spacing issues in the {pipelines-title} Web UI. With this update, the console plugin styling is updated to use the appropriate PatternFly equivalent classes, ensuring consistent alignment and visual integration with the current OpenShift console design standards.

* Before this update, Web UI links in `PipelineRun` incorrectly pointed to the deprecated `v1beta1` {pipelines-title} APIs instead of the current `v1` APIs. With this update, the links point to the appropriate `v1` APIs.

> * Before this update, {pipelines-title} and {tekton-results} incorrectly displayed `TaskRun` resources from previous `PipelineRun` that shared the same name. This led to confusion about which `TaskRun` resources were associated with the current execution. With this update, Tekton Results now correctly isolates and displays only the `TaskRun` resources associated with the current `PipelineRun`, preventing the mixing of archived and active execution data.

Author: Fixed.

> * Before this update, end-to-end (E2E) tests encountered instability due to GitOps comments being incorrectly associated with cancelled pipelinerun. This behavior caused intermittent test failures and reduced reliability in CI/CD pipelines. With this update, GitOps comments are no longer mixed with cancelled runs, resulting in a stable and predictable E2E test execution.

Author: Fixed.

> * Before this update, the Pipelines Dynamic plugin used shared proxy endpoints that were associated with known security vulnerabilities (CVEs). This posed potential risks and reduced the reliability of the plugin. With this update, the Pipelines Dynamic plugin uses dedicated API endpoints, mitigating the identified security issues and enhancing overall reliability and security posture.

Reviewer 4: Eliska is right.

> * Before this update, the `tekton-caches` `tarit` utility did not preserve file permissions when compressing cached directories. This led to issues with executable files and scripts after decompression, particularly when artifacts were used by different users or within SELinux-enforcing base images. With this update, file permissions are now properly retained during cache operations, ensuring compatibility and correct behavior across different user contexts.

Author: Hey @eromanova97, the content provided by the developers in the Google doc is mainly edited by using AI-related tools, and hence, I am faced with the job of trying to change the tone of the sentence and trying to use simple language. This problem can be resolved at the root.

> * Before this update, when a `TaskRun` failed due to `ImagePullBackOff` errors, the `PipelineRun` log snippet displayed unclear messages, such as “pods not found,” after switching between tabs in the {pipelines-title} Web UI. With this update, clear error messages, such as `TaskRunImagePullFailed` or `failing to pull image`, are now displayed to improve the troubleshooting experience.

Author: Fixed.

> * Before this update, certain elements within the {pipelines-title} Start interface, such as `deployment-name`, `Hr`,  `Min`, and `Sec`, were presented in English irrespective of the user's selected regional settings. With this update, comprehensive localization is now incorporated for all interface elements, ensuring these elements are displayed according to the user's regional settings.

Author: Fixed.

> * Before this update, the Tekton pruner job encountered `ImagePullBackOff` errors during Helm-based installation. This issue was caused by missing `SHA256 digests` in image tags. With this update, image tag handling includes the required `SHA256 digests` for reliable image pulling.

Author: Fixed.

* Before this update, the `openshift-pipelines-console-plugin-rhel8-container` image was affected by a vulnerability in the `serve-static` package caused by improper input sanitization. With this update, the vulnerability is resolved by updating the affected packages.

* Before this update, the `openshift-pipelines-console-plugin-rhel8-container` image was affected by a Denial of Service (DoS) vulnerability in the `body-parser` package. With this update, the vulnerability is fixed through package updates.

* Before this update, the {pac} controller could crash with an “index out of range” error during push events from Bitbucket Data Center. This behavior occurred when the changes array in the event payload was empty. With this update, {pac} now handles empty changes arrays gracefully, preventing the controller from crashing.

* Before this update, adding labels to pull requests would unintentionally trigger a `PipelineRun`. With this update, this issue is resolved, preventing unintended `PipelineRun` triggers from label additions.

> * Before this update, closing a pull request would cancel ongoing `PipelineRun` even if the `cancel-in-progress` annotation was not set. With this update, `PipelineRun` are now only canceled on pull request closure when explicitly configured using the `cancel-in-progress` annotation.

Author: Fixed.

* Before this update, the GitLab integration in {pac} encountered API call failures caused by an incorrect API URL. With this update, this issue is fixed by introducing URL validation, which prevents such misconfigurations and ensures successful API communication.

> * Before this update, {pac} did not cancel a `PipelineRun` created with the `generateName` field, even when the `cancel-in-progress` annotation was set. With this update, {pac} correctly cancels an in-progress `PipelineRun` having the `generateName` field.

Author: Fixed.

* Before this update, when provenance was configured in GitLab, {pac} retrieved an incorrect PipelineRun template from the Git repository. With this update, {pac} correctly identifies and retrieves the intended template in GitLab provenance setups.

> * Before this update, the `/ok-to-test` GitOps command triggered a `PipelineRun` when used in a push commit comment. With this update, the `/ok-to-test` command no longer triggers a `PipelineRun` when used outside of pull request contexts.

Author: Fixed.

* Before this update, `TaskRuns` and `PipelineRun` failed when referencing `StepActions` by the Artifact Hub resolver, returning the error `kind param must be task or pipeline`, as `StepAction` was not recognized as a valid resource type. With this update, the Artifact Hub resolver supports `StepAction` references, allowing users to include remote step actions in their tasks and pipelines.

> * Before this update, `PipelineRun` failed with the error `failed to create subPath directory for volumeMount`, even though Kubernetes would eventually recover and create the required pod. This led to unnecessary `PipelineRun` failures and a poor user experience, often requiring manual restarts. With this update, `PipelineRun` implements a grace period and retry mechanism for `subPath` directory creation errors. This allows Kubernetes time to resolve the issue automatically, reducing false failures and improving reliability.

Reviewer 4: We do use OpenShift not Kubernetes downstream, so replace Kubernetes with {OCP}

* Before this update, the {tekton-results} API server encountered an error when a log query was made for a non-existent `TaskRun` or `PipelineRun`. With this update, this issue is fixed as such queries are now handled gracefully without causing an error.

* Before this update, users had to wrap the CLI in separate Tasks to back up or restore build caches. With this update, Tekton `StepActions` can directly use `fetch` and `upload`, allowing cache operations to be performed with a single-line step inside any Task or Pipeline.

> * Before this update, pushing a cache to a registry with a self-signed certificate failed due to TLS errors. With this update, the new `--insecure flag`, available in  the CLI and Task, enables those pushes making it easier to work with air-gapped development clusters and local registries.

Author: Fixed.

* With this update, the `fetch` command is improved to automatically create the destination folder if it does not exist in a new workspace. Previously, the command would fail in such cases, requiring you to create a directory manually.

* With this update, registry authentication is no longer limited to the `/tekton/home/.docker/config.json` default path. You can now mount any Docker configuration file and specify its location by using the `dockerConfig` parameter in your `Task` resource.
+
.Example for enabling the `dockerConfig` parameter in a task
[source,yaml]
----
apiVersion: tekton.dev/v1 
kind: Task 
metadata:
  name: task
spec: 
params:
- name: dockerConfig
  value: /workspace/creds/config.json
# ...
----

[id="tekton-chains-new-features-1-19_{context}"]
=== {tekton-chains}

* With this update, the {tekton-chains} controller uses `StatefulSet` ordinals to improve high availability and workload distribution as an alternative to the leader election mechanism.
+
--
.Example of enabling the StatefulSet ordinals for the the `Chains` controller
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonChains
metadata:
  name: chains
spec:

>   chains:

Reviewer 5: Should be chain instead of chains

spec:
  chain:
    performance:
      disable-ha: false

Author: @manthinasai - Please check the content, it is fixed now.
